### ISSUE TITLE
Refactor spacing code

### ISSUE THREAD

**Issue by eroux** [gregorio-project/gregorio#63]
Spacing code and algorithm is currently extremely complex to understand, maintain and improve, Here are a few ideas to simplify it:

 * there should be mainly one space macro for space between notes and one space macro for space between words
 * for normal syllables, the algorithm is simply to go left as much as possible, with two constraints: 
   * the notes of the different syllables are separated by a certain space
   * letters are separated by a certain space if from different words
 * for syllables empty or with only a bar (with or without text), the algorithm is slightly more complex:
   * previous notes, bar and next notes are all separated by at least a certain space, and bar is centered (in case of syllable with no bar, the constraints are only between the notes of the previous and next syllables
   * previous text, text and next text are separated by at least a certain space, and the text is centered (in case of syllable with no text, the constraints are only between the previous and next text)

**Comment by rpspringuel** [comment]
I'm finding that my work on #50 is involving some significant refactoring of the spacing code.  I don't know that it will address the issues you named above, but it might make implementing them a bit easier.  Unfortunately, I made too many changes at once and am having some problems tracking down bugs at the moment, but I have the concept that I want in my head now.  I'm probably going to try reverting and making the changes in a more piecemeal approach in order to better identify the source of the bugs.  However, any significant work on this should probably wait until after I get done with #50.

**Comment by eroux** [comment]
The main changes won't really be on space definition, but mostly in \gresyllable, \grebarsyllabe, \greendofword and \greendofsyllable, do you think we can work on it in parallel? In case you think it's dangerous, I can wait for you to finish #50, I'm not in a hurry...

Also, I was thinking about making a few lua helpers for spacing algorithm, this will make things way more graphical... for instance replacing

    \b = \a
    \multiply\b by 2
    \advance\b by \c

by

    b = 2 * a + c

improves readability and maintainability by orders of magnitude...

What do you think?

**Comment by rpspringuel** [comment]
The biggest changes in those functions that I'm making would be to change the names of the distances from `\gredistance` to `\gre@distance` and to change the names of functions which calculate distances (and have no other purpose) to `\gre@calculate@distance`.  The other big change that might affect you is to aggregate all distance declarations into `gregoriotex-spaces.tex`.  As a result you're probably safe working on those.  I'm trying the piecemeal changes now and these are the first two things I'm doing (the aggregation is already done).  I should have that component done by the end of the day.

The helpers would definitely be a huge improvement.

**Comment by rpspringuel** [comment]
Just finished those two steps.  See f2883987f49c257bb7256fe8227ff87203415df8 for code details.

**Comment by eroux** [comment]
Ok, wonderful! I'll work on this soon!

**Comment by eroux** [comment]
@rpspringuel do you think you could merge your work soon? If you're not confident about it, you can merge it in a branch (not master), but it would be useful if I could take it as a basis... Also, do not hesitate to create branches on the main gregorio repo, we're not an army of developers needing a workflow with many different repos, so I believe there're enough room for everyone on the main repo if we work on separate branches!

**Comment by henryso** [comment]
Is this still open?

**Comment by rpspringuel** [comment]
It is, but @eroux is looking at some possible changes to LuaTeX itself to make what he wants to do easier.  This is waiting in the wings as a result.

**Comment by henryso** [comment]
There are two outstanding commits in `fix-63`.  Are these part of the LuaTeX changes?

**Comment by eroux** [comment]
No, these are just some code reorganization for future changes, but it's not tested enough, I'll work on it later...

**Comment by eroux** [comment]
Closing this, as things took a different road
